projects/optic/src/commands/diff/diff-all.ts

```diff
@@ -0,0 +1,447 @@
+
+const helpText = `
+Example usage:
+  Diff all specs with \`x-optic-url\` in the current repo against HEAD~1
+  $ optic diff-all
+
+  Diff all specs with \`x-optic-url\` in the current repo from main to feature/1
+  $ optic diff-all --compare-from main --compare-to feature/1
+
+  Diff all specs with a ruleset, run checks and open up in a web browser
+  $ optic diff-all --ruleset @org/example-ruleset --web --check
+  `;
+
+export const registerDiffAll = (cli: Command, config: OpticCliConfig) => {
+  cli
+    .command('diff-all')
+    .configureHelp({
+      commandUsage: usage,
+    })
+    .addHelpText('after', helpText)
+    .description('Run a diff on all specs with `x-optic-url`')
+    .option(
+      '--compare-to <compare-to>',
+      'the head ref to compare against. Defaults to the current working directory'
+    )
+    .option(
+      '--compare-from <compare-from>',
+      'the base ref to compare against. Defaults to HEAD~1',
+      'HEAD~1'
+    )
+    .option(
+      '--ruleset <ruleset>',
+      'run comparison with a locally defined ruleset, if not set, looks for the ruleset on the [x-optic-ruleset] key on the spec, and then the optic.dev.yml file.'
+    )
+    .option('--check', 'enable checks', false)
+    .option('--web', 'view the diff in the optic changelog web view', false)
+    .option('--json', 'output as json', false)
+    .action(wrapActionHandlerWithSentry(getDiffAllAction(config)));
+};
+
+type DiffAllActionOptions = {
+  compareTo?: string;
+  compareFrom: string;
+  ruleset?: string;
+  check: boolean;
+  web: boolean;
+  json: boolean;
+};
+
+// Match up the to and from candidates
+// This will return the comparisons we can try to run
+function matchCandidates(
+  from: {
+    ref: string;
+    paths: string[];
+  },
+  to: { ref?: string; paths: string[] }
+): Map<
+  string,
+  {
+    from?: string;
+    to?: string;
+  }
+> {
+  const results = new Map<string, { from?: string; to?: string }>();
+  for (const path of from.paths) {
+    const strippedPath = path.replace(`${from.ref}:`, '');
+    results.set(strippedPath, {
+      from: path,
+    });
+  }
+
+  for (const path of to.paths) {
+    const strippedPath = to.ref ? path.replace(`${to.ref}:`, '') : path;
+    const maybePathObject = results.get(strippedPath);
+    if (maybePathObject) {
+      maybePathObject.to = path;
+    } else {
+      results.set(strippedPath, {
+        to: path,
+      });
+    }
+  }
+  return results;
+}
+
+async function computeAll(
+  candidatesMap: ReturnType<typeof matchCandidates>,
+  config: OpticCliConfig,
+  options: DiffAllActionOptions
+): Promise<{
+  warnings: Warnings;
+  results: Result[];
+}> {
+  const warnings: Warnings = {
+    missingOpticUrl: [],
+    unparseableFromSpec: [],
+    unparseableToSpec: [],
+  };
+
+  const results: Result[] = [];
+
+  for await (const [_, candidate] of candidatesMap) {
+    // try load both from + to spec
+    let fromParseResults: ParseResult;
```

Comment:
Here we'll only compare if:
- `to` spec is parseable (with strict validation)
- `from` spec is parseable (with not strict validation)
- `to` spec has the x-optic-url (we've opted in) OR from spec has `x-optic-url` AND to spec is empty (i.e. we've deleted the `to` spec)